### PR TITLE
Issue/#316 add epm to validator

### LINKED ISSUE #324: Validation script creates new smac3-output... folder
This folder contains nothing but a copy of the scenario from the directory I wanted to validate.
I think the script should not create a new smac output folder (especially an empty one) as the --output flag determines where the output will be created

### PR DESCRIPTION
**Pull request by shukon**
Fixes #324 (unnecessary output-dirs).
Adds the option to validate using an EPM.
The option to evaluate configurations over `time` (formerly at timesteps 2^-4, 2^-3, 2^-2, ... 2^0, 2^1, ...) is replaced by the options `wallclock_time` and `cpu_time`. Also issue #319 should be fixed, now using the maximum time found in the trajectory and adding configs backwards (at timesteps max_time/2^0, max_time/2^1, ..., 0).

### CODE REVIEW

File: test/test_utils/test_validate.py [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 from smac.tae.execute_ta_run import StatusType
+from smac.tae.execute_ta_run_old import ExecuteTARunOld
 from smac.runhistory.runhistory import RunHistory
 from smac.stats.stats import Stats
 from smac.optimizer.objective import average_cost
 from smac.utils.io.traj_logging import TrajLogger
-from smac.utils.validate import Validator
+from smac.utils.validate import Validator, Run
+
+if sys.version_info[0] == 2:
```
Comment (mlindauer) [inline test/test_utils/test_validate.py]: why is this important? We don't support Python 2 anymore.

File: scripts/smac-validate.py [hunk 2]
```diff
@@ -75,15 +79,29 @@
...
-    validator.validate(config_mode=args_.configs,
-                       instance_mode=args_.instances,
-                       repetitions=args_.repetitions,
-                       n_jobs=args_.n_jobs,
-                       runhistory=args_.runhistory,
-                       tae=tae)
+
+    # Load runhistory
+    if args_.runhistory:
+        runhistory = RunHistory(average_cost)
```
Comment (mlindauer) [inline scripts/smac-validate.py]: Can we use only a single runhistory or could we use also use all runhistories from independent runs?

File: smac/utils/validate.py [hunk 3]
```diff
@@ -189,43 +203,137 @@ def _validate_parallel(self, tae, runs, n_jobs, backend):
...
+        # Predict desired runs
+        runs = self.get_runs(config_mode, instance_mode, repetitions, runhistory)
+        try:
+            feature_array_size = len(self.scen.cs.get_hyperparameters()) + self.scen.feature_array.shape[1]
+        except AttributeError:
+            feature_array_size = len(self.scen.cs.get_hyperparameters())
+        X_pred = np.empty((len(runs), feature_array_size))
+        for idx, run in enumerate(runs):
+            if hasattr(self.scen, "feature_dict") and run.inst != None:
+                X_pred[idx] = np.hstack([run.config.get_array(),
```
Comment (mlindauer) [inline smac/utils/validate.py]: I think you have to use `from smac.configspace import convert_configurations_to_array` such that the inactive parameters are also correctly imputed

File: smac/utils/validate.py [hunk 4]
```diff
@@ -189,43 +203,137 @@ def _validate_parallel(self, tae, runs, n_jobs, backend):
...
+                                         self.scen.feature_dict[run.inst]])
+            else:
+                X_pred[idx] = run.config.get_array()
+        self.logger.debug("Predicting desired %d runs, data has shape %s and "
+                          "NaNs are present: %s (if so, will be zeroed)",
+                          len(runs), str(X_pred.shape), str(np.isnan(X_pred).any()))
+
+        X_pred = np.nan_to_num(X_pred)  # Otherwise segfault in rfr
```
Comment (mlindauer) [inline smac/utils/validate.py]: that's not necessary if you use convert_configurations_to_array beforehand.

File: scripts/smac-validate.py (hunk 2 above)
Comment (shukon) [reply scripts/smac-validate.py]: Now we can use multiple ones by using `--runhistory path_to_rh_1 path_to_rh_2 path_to_rh_3` and so on.

File: test/test_utils/test_validate.py (hunk 1 above)
Comment (shukon) [reply test/test_utils/test_validate.py]: True.

File: smac/utils/validate.py (hunk 3 above)
Comment (shukon) [reply smac/utils/validate.py]: Thanks. Changed.

**Comment by shukon** [comment]
@mlindauer Can you review again? I fixed your requests and also added #319 and minor other fixes.

**Comment by shukon** [comment]
@mlindauer does the current version fulfill your requested changes?

**Comment by mlindauer** [comment]
Yes, thanks.

**Comment by mlindauer** [comment]
@mfeurer do you also want to review this pr or can we simply merge it?

**Pull request merged**